server/groups/keycloak.go

```diff
@@ -502,80 +562,102 @@ func (k *KeycloakClient) HandleSAMLLogin(c *plugin.Context, user *mmModel.User,
+					return errors.New("failed to get teams for removal")
+				}
 			}
-			if err := k.getSyncableChannelsForRemoval(groupID, proposedChannelsToLeave); err != nil {
-				k.PluginAPI.Log.Error("Failed to get channels for removal",
-					"group_id", groupID,
-					"error", err)
+			if err = k.getSyncableChannelsForRemoval(groupID, proposedChannelsToLeave); err != nil {
+				if k.LockOutUsersOnRemovalFailure {
+					return errors.New("failed to get channels for removal")
+				}
 			}
 		}
 	}
 
 	// Process all active groups (both remaining and newly added)
 	if len(activeGroups) > 0 {
 		for _, groupID := range activeGroups {
-			if err := k.getSyncableTeamsForAddition(groupID, finalTeamsToJoin); err != nil {
-				k.PluginAPI.Log.Error("Failed to get teams for addition",
-					"group_id", groupID,
-					"error", err)
-			}
-			if err := k.getSyncableChannelsForAddition(groupID, finalChannelsToJoin); err != nil {
-				k.PluginAPI.Log.Error("Failed to get channels for removal",
-					"group_id", groupID,
-					"error", err)
-			}
+			_ = k.getSyncableTeamsForAddition(groupID, finalTeamsToJoin)
+			_ = k.getSyncableChannelsForAddition(groupID, finalChannelsToJoin)
 		}
 	}
 
-	finalTeamsToLeave := make(map[string]bool)
+	finalTeamsToLeave := make(map[string]mmModel.GroupSyncable)
 	// Loop over teams to leave and check if it's in the list of teams to join.
 	// If it's in the teams to join, we don't need to remove the team membership.
-	for teamID := range proposedTeamsToLeave {
+	for teamID, groupSyncable := range proposedTeamsToLeave {
 		if _, exists := finalTeamsToJoin[teamID]; !exists {
-			finalTeamsToLeave[teamID] = true
+			finalTeamsToLeave[teamID] = groupSyncable
 		}
 	}
 
-	finalChannelsToLeave := make(map[string]bool)
+	finalChannelsToLeave := make(map[string]mmModel.GroupSyncable)
 	// Loop over channels to leave and check if it's in the list of channels to join.
 	// If it's in the channels to join, we don't need to remove the channel membership.
-	for channelID := range proposedChannelsToLeave {
+	for channelID, groupSyncable := range proposedChannelsToLeave {
 		if _, exists := finalChannelsToJoin[channelID]; !exists {
-			finalChannelsToLeave[channelID] = true
+			finalChannelsToLeave[channelID] = groupSyncable
+		}
+	}
+	err = k.removeUserFromChannels(finalChannelsToLeave, user)
+	if err != nil {
+		if k.LockOutUsersOnRemovalFailure {
+			return errors.New("failed to remove user from channel")
+		}
+	}
+	err = k.removeUserFromTeams(finalTeamsToLeave, user)
+	if err != nil {
+		if k.LockOutUsersOnRemovalFailure {
+			return errors.New("failed to remove user from team")
+		}
+	}
+	err = k.RemoveUserFromGroups(groupsForRemoval, user)
+	if err != nil {
+		if k.LockOutUsersOnRemovalFailure {
+			return errors.New("failed to remove user from group")
+		}
+	}
+	err = k.AddUserToGroups(activeGroups, user)
+	if err != nil {
+		if k.LockOutUsersOnRemovalFailure {
+			return errors.New("failed to add user to group")
```

Comment:
I know I said I'd only do this for removals but I wonder if we should do it for group additions as well? Up for debate